File: app-tailwind-classes/src/003-Tailwind/MonacoCssEdditor.cpp
#include "003-Tailwind/MonacoCssEdditor.h"
#include <Wt/WText.h>
#include <Wt/WApplication.h>
#include <Wt/WPushButton.h>

MonacoCssEdditor::MonacoCssEdditor()
    : js_signal_text_changed_(this, "cssEdditorTextChanged")
{

    setStyleClass("min-w-[500px] min-h-[500px]");

    js_signal_text_changed_.connect(this, &MonacoCssEdditor::cssEdditorTextChanged);
    // doJavaScript("require.config({ paths: { 'vs': 'https://unpkg.com/monaco-editor@0.34.1/min/vs' } });");
    std::string js_member = "";
    js_member += "require(['vs/editor/editor.main'], function() {";
    js_member += "  window.cssEditorInstance = monaco.editor.create(document.getElementById('" + id() + "'), {";
    js_member += "    language: 'css',";
    js_member += "  });";
    js_member += "  window.cssEditorInstance.onDidChangeModelContent(function() {";
    js_member += js_signal_text_changed_.createCall({ " window.cssEditorInstance.getValue()" });
    // js_member += "    Wt.emit("+id()+", 'cssEdditorTextChanged', window.cssEditorInstance.getValue());";
    js_member += "  });";
    js_member += "});";   

    // setJavaScriptMember("initialize", js_member);
    // setJavaScriptMember("initializeCssEditor("+id()+")", js_member);
    // Wt::WApplication::instance()->doJavaScript("initializeCssEditor('" + id() + "', '"+"alkjdshfaijksdhf i"+"');", true);
    std::string member = "initializeCssEditor('" + id() + "', '"+"alkjdshfaijksdhf i"+"');";
    setJavaScriptMember("cssEdditorTextChanged", member);
    // Wt::WApplication::instance()->doJavaScript("updateCssEditorValue('body { background-color: red; }')", true);
}

void MonacoCssEdditor::cssEdditorTextChanged(const std::string text)
{
    std::cout << "\nText changed: " << text << "\n";
}

void MonacoCssEdditor::setCssEdditorText(std::string text)
{
    // doJavaScript("window.cssEditorInstance.setValue('" + text + "');");
    doJavaScript("updateCssEditorValue('" + text + "');");
}


void MonacoCssEdditor::setDarkTheme(bool dark)
{
    // doJavaScript("toggleEditorTheme()");
    if(dark)
        doJavaScript("if(monaco.editor) monaco.editor.setTheme('vs-dark');");
    else
        doJavaScript("if(monaco.editor) monaco.editor.setTheme('vs');");
}

File: static/js/monaco-css-edditor.js
let edditor_css; // Declare the global variable
let current_text = ""; // Initialize current_text variable
require.config({ paths: { 'vs': 'https://unpkg.com/monaco-editor@0.34.1/min/vs' } });

function initializeCssEditor(elementId, edditor_content) {
  require(['vs/editor/editor.main'], function () {
    edditor_css = monaco.editor.create(document.getElementById(elementId), { // Assign to global variable
      language: 'css',
      value: edditor_content,
      theme: 'vs-dark',
      wordWrap: 'on',
      lineNumbers: 'on',
      tabSize: 4, // Set the tab size
      insertSpaces: false, // Use tabs instead of spaces
      detectIndentation: false, // Disable automatic indentation detection
      trimAutoWhitespace: false, // Prevent trimming of whitespace
      lineEnding: '\n' // Ensure consistent line endings
    });

    // Event listener for content changes
    edditor_css.onDidChangeModelContent(function (event) {
      if(current_text !== edditor_css.getValue()) { // Check if the text has changed`
        current_text = edditor_css.getValue(); // Update current_text with the new value
        Wt.emit(elementId, 'cssEdditorTextChanged', edditor_css.getValue());
      }
    });
    // Override the Ctrl+S command
    edditor_css.getDomNode().addEventListener('keydown', function(e) {
      if ((e.ctrlKey || e.metaKey) && e.key === 's') {
        e.preventDefault(); // Prevent the browser's default save dialog
        // Wt.emit(elementId, 'cssEdditorSaveText');
      }
    });
  });
}


function updateCssEditorValue(newValue) {
  if (edditor_css) {
    current_text = newValue; // Store the current text before updating
    edditor_css.setValue(newValue);
  } else {
    console.error("Editor instance is not initialized yet.");
  }
}

// function toggleEditorTheme() {
//   const currentTheme = window.cssEditorInstance._themeService._theme.themeName;
//   console.log("Current theme:", currentTheme);
//   if (window.cssEditorInstance) {
//     // const newTheme = currentTheme === 'vs-dark' ? 'vs' : 'vs-dark';
//     // window.cssEditorInstance.setTheme(newTheme);
//   } else {
//     console.error("Editor instance is not initialized yet.");
//   }
// }